### PR TITLE
EZP-31129: Added optional parameter to unlinkURL to better delete unused links

### PR DESCRIPTION
**Pull request by mateuszdebinski**
| Question           | Answer
| ------------------ | ------------------
| **JIRA issue**     | [EZP-31129](https://jira.ez.no/browse/EZP-31129)
| **Bug/Improvement**| yes
| **New feature**    | no
| **Target version** | 7.5
| **BC breaks**      | no
| **Tests pass**     | yes
| **Doc needed**     | no

Depended on: https://github.com/ezsystems/ezplatform-richtext/pull/109

**TODO**:
- [x] Implement feature / fix a bug.
- [x] Implement tests.
- [x] Fix new code according to Coding Standards (`$ composer fix-cs`).
- [x] Ask for Code Review.

### CODE REVIEW

File: eZ/Publish/Core/FieldType/Url/UrlStorage/Gateway.php [hunk 1]
```diff
@@ -61,6 +61,7 @@ abstract class Gateway extends StorageGateway
      *
      * @param int|string $fieldId
      * @param int $versionNo
+     * @param array|null $excludeUrlIds
      */
-    abstract public function unlinkUrl($fieldId, $versionNo);
+    abstract public function unlinkUrl($fieldId, $versionNo, $excludeUrlIds);
```
Comment (alongosz) [inline eZ/Publish/Core/FieldType/Url/UrlStorage/Gateway.php]: optional, also making it null is pointless due to the usage
```suggestion
    abstract public function unlinkUrl($fieldId, $versionNo, array $excludeUrlIds = []);
```

File: eZ/Publish/Core/FieldType/Url/UrlStorage/Gateway/DoctrineStorage.php [hunk 2]
```diff
@@ -162,8 +163,9 @@ public function linkUrl($urlId, $fieldId, $versionNo)
      *
      * @param int $fieldId
      * @param int $versionNo
+     * @param array|null $excludeUrlIds
      */
-    public function unlinkUrl($fieldId, $versionNo)
+    public function unlinkUrl($fieldId, $versionNo, $excludeUrlIds = null)
```
Comment (alongosz) [inline eZ/Publish/Core/FieldType/Url/UrlStorage/Gateway/DoctrineStorage.php]:
```suggestion
    public function unlinkUrl($fieldId, $versionNo, array $excludeUrlIds = [])
```

File: eZ/Publish/Core/FieldType/Url/UrlStorage/Gateway/DoctrineStorage.php [hunk 3]
```diff
@@ -181,10 +183,20 @@ public function unlinkUrl($fieldId, $versionNo)
...
         ;
 
+        if (empty($excludeUrlIds) === false) {
+            $deleteQuery
+                ->andWhere(
+                    $deleteQuery->expr()->notIn(
+                        'url_id',
+                        ':url_ids'
+                    )
+                )->setParameter(':url_ids', implode(', ', $excludeUrlIds), ParameterType::STRING);
```
Comment (alongosz) [inline eZ/Publish/Core/FieldType/Url/UrlStorage/Gateway/DoctrineStorage.php]:
```suggestion
                )
                ->setParameter('url_ids', $excludedUrlIds, Connection::PARAM_INT_ARRAY);
```

File: eZ/Publish/Core/FieldType/Url/UrlStorage/Gateway.php [hunk 4]
```diff
@@ -61,6 +61,7 @@ abstract class Gateway extends StorageGateway
      *
      * @param int|string $fieldId
      * @param int $versionNo
+     * @param array|null $excludeUrlIds
```
Comment (alongosz) [inline eZ/Publish/Core/FieldType/Url/UrlStorage/Gateway.php]:
```suggestion
     * @param int[] $excludeUrlIds
```

File: eZ/Publish/Core/FieldType/Url/UrlStorage/Gateway/DoctrineStorage.php [hunk 5]
```diff
@@ -162,8 +163,9 @@ public function linkUrl($urlId, $fieldId, $versionNo)
      *
      * @param int $fieldId
      * @param int $versionNo
+     * @param array|null $excludeUrlIds
```
Comment (alongosz) [inline eZ/Publish/Core/FieldType/Url/UrlStorage/Gateway/DoctrineStorage.php]:
```suggestion
     * @param int[] $excludeUrlIds
```

**Comment by mateuszdebinski** [comment]
> Please add test coverage!

Link to test in RichText: https://github.com/ezsystems/ezplatform-richtext/blob/901f22ca18b2a20601009373893925bc7a8d14c0/tests/integration/eZ/API/RichTextFieldTypeIntegrationTest.php#L639

**Comment by lserwatka** [comment]
@mateuszdebinski could you merge it to ezplatform-kernel/master?

**Pull request merged**